haskell.compiler.ghc{8107{,Binary},928}: drop

emilazy opened the pull request:
What can I say? I felt left out of all the fun…

This probably has issues since I don’t really know what I’m doing Haskell-package‐set‐wise, but I felt more like doing this than building another GHC 9.2 to validate https://github.com/NixOS/nixpkgs/pull/440271. It’s based on a rebased/fixed up version of https://github.com/NixOS/nixpkgs/pull/381265, which should of course land first.
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy committed:
haskell.compiler.ghc{928,963,967,9101,9102,HEAD}: drop AArch32 bootstrap

It’s questionable whether or not this worked – 32‐bit ARM in
Nixpkgs is already flaky even when cross‐compiling, and I find
it dubious whether much 32‐bit ARM hardware can build modern
GHCs in a bearable amount of time and memory – and figuring out
cross‐compilation of bootstrap GHCs will be a more sustainable way
to keep it working if anyone cares enough to.

Also, rearrange the GHC‐related release notes to be in order of
most likely to matter to anyone.

emilazy committed:
haskell.compiler.ghc902: add missing alias

Fixes: f8560ade011b8f63a002b80152f81e624c743336

emilazy committed:
haskell.compiler.ghc8107{,Binary}: drop

Also, rearrange the GHC‐related release notes to be in order of
most likely to matter to anyone.

Co-authored-by: Wolfgang Walther <walther@technowledgy.de>

emilazy committed:
haskell.compiler.ghc928: drop

emilazy commented:
“GHC 9.4.8 does not have a binary” would be one of those issues, especially in light of https://github.com/NixOS/nixpkgs/pull/440410. I guess bootstrapping with a newer GHC than the one you’re building won’t work, so… they’d have to use 9.0.2 (if that works?) or else the source‐based 9.4.8?

wolfgangwalther reviewed (dismissed):
Comparing notes with my local branch on the same topic, see comments. I also have some changes in the docs to change `haskell.packages.ghc92` to `haskellPackages` in various examples, as well as a mention of `ghcName = "ghc92"`, which I changed to `ghc910`, might as well change it to `ghc912` to last longer.

I think we should back out the drop of `ghc924Binary` for now to make progress on the remaining stuff.

Also needs a rebase, ofc.

emilazy commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -26,7 +26,7 @@
 - GHC 8.6.5 and its package set have been removed. It was only used to bootstrap GHC for powerpc64le, but this was probably broken anyway.
 
-- GHC 9.0.2 and its package set have been removed.
+- GHC 8.10 and 9.0.2, and their package sets, have been removed.
```
Done.

emilazy commented on pkgs/top-level/release-haskell.nix:
```diff
@@ -591,12 +581,6 @@ let
       titlecase = released;
-      ghc-api-compat = [
-        compilerNames.ghc8107
-      ];
```
I should have known my attempt at getting some bonus removal fun would lead me to having to actually interact with the Haskell package stuff that frightens and confuses me.

Should I also be doing something about the `integer-simple` Hackage package that is, uh, marked broken? (Was any of that stuff *actually working*?)

emilazy replied:
Why does that script want to build Nix 2.28…

Worked around with `nixpkgs=flake:nixpkgs/nixpkgs-unstable`; I guess you don’t want to use the CI pin because `haskell-updates` can bump the relevant tooling?

But then:

```
shion:~/Developer/nixpkgs  1  (no description set)
❭ maintainers/scripts/haskell/regenerate-hackage-packages.sh --fast
fetching path input 'path:/nix/store/ngghfaknb50axlvdnxclr6mwgkp9ym3b-source'
Obtaining Hackage data …
Generating compiler configuration …
Running hackage2nix to regenerate pkgs/development/haskell-modules/hackage-packages.nix …
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pty-mcp-server/0.0.5.0/pty-mcp-server.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-ui-response/0.0.4.0/pms-ui-response.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-ui-request/0.0.4.0/pms-ui-request.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-ui-notification/0.0.3.0/pms-ui-notification.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-infrastructure/0.0.4.0/pms-infrastructure.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-infra-watch/0.0.3.0/pms-infra-watch.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-infra-socket/0.0.1.0/pms-infra-socket.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-infra-cmdrun/0.0.2.0/pms-infra-cmdrun.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-infra-procspawn/0.0.1.0/pms-infra-procspawn.cabal)
hackage2nix: user error (cannot parse cabal file /nix/store/xk82157yf0986i80f19wvmwjq45jvgvf-unpacked-cabal-hashes/pms-domain-service/0.0.4.0/pms-domain-service.cabal)
hackage2nix: thread blocked indefinitely in an MVar operation
```

Is this a Darwin thing?

I’m going to push out the changes to the `.yml` but not actually update the `.nix` file until I figure out what’s going on here.

emilazy committed:
haskell.compiler.ghc948: remove obsolete bootstrap hack

We do, in fact, have a binary package for this now, and we do, in fact,
use it.

emilazy commented on pkgs/top-level/release-haskell.nix:
```diff
@@ -591,12 +581,6 @@ let
-      ghc-api-compat = [
-        compilerNames.ghc8107
-      ];
-      ghc-bignum = [
-        compilerNames.ghc8107
-      ];
```
Done.

emilazy commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -26,7 +26,7 @@
 - GHC 8.6.5 and its package set have been removed. It was only used to bootstrap GHC for powerpc64le, but this was probably broken anyway.
 
-- GHC 8.10 and 9.0.2, and their package sets, have been removed.
+- GHC 8.10, 9.2, and 9.0.2, and their package sets, have been removed.
```
Done. Whoops.

emilazy commented on pkgs/development/compilers/ghc/common-make-native-bignum.nix:
```diff
@@ -506,7 +472,9 @@ stdenv.mkDerivation (
     # HACK: allow bootstrapping with GHC 8.10 which works fine, as we don't have
     # binary 9.0 packaged. Bootstrapping with 9.2 is broken without hadrian.
-    + lib.optionalString (lib.versions.majorMinor version == "9.4") ''
+    #
+    # TODO: Remove this, as we don’t do this any more.
+    + ''
```
I guess not a meaningful number of them. Done.

emilazy commented:
> Can do, and also run treefmt while at it.

Thanks. (Sorry, I need to set up an alias to `jj fix` my commits before they get pushed; relying on CI for reasonable eval timing makes my Nixpkgs workflow sloppy…)

emilazy commented on pkgs/development/haskell-modules/configuration-hackage2nix/main.yaml:
```diff
@@ -114,7 +114,6 @@ extra-packages:
   - text-metrics < 0.3.3                # 2025-02-08: >= 0.3.3 uses GHC2021
   - versions < 6                        # 2024-04-22: required by spago-0.21
   - weeder == 2.2.*                     # 2022-02-21: preserve for GHC 8.10.7
-  - weeder == 2.3.*                     # 2022-05-31: preserve for GHC 9.0.2
```
Er, not the follow‐up, since that’s 9.2‐related.

I still feel like we shouldn’t maintain an entire 9.0 package set just because it’s load‐bearing for bootstrap, but that’s a separate matter. Like, I feel confident about `subopt`, but I’d still rather it only be used for a defined path to a source build rather than being treated as a first‐class thing for end users. The lack of `throw` aliases made me assume that that was the intention of the binary packages in general.

emilazy replied:
You probably want to drop the 9.2 bits for me to move into the follow‐up PR too, though. (Or I can just do it if you’d prefer.)

emilazy replied:
Sounds good to me. Sorry for making more work for you here!

emilazy commented:
> FWIW, my preference is to land this before #440774 (or #440273 if the unpins get moved there),

(Er, I meant https://github.com/NixOS/nixpkgs/pull/440271 here. But now there should be no textual conflict between the two, as the LLVM pinning changes will be moved to https://github.com/NixOS/nixpkgs/pull/440273. So I’d still like to see this land soon so I can rebase https://github.com/NixOS/nixpkgs/pull/440273, or else know that it won’t land soon. Of course, I’ll also rebase https://github.com/NixOS/nixpkgs/pull/440774 once the `master`‐targeting churn is merged and makes its way to `haskell-updates`.)

emilazy commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -24,11 +24,11 @@
-- GHC 9.0 and its package set have been removed.
+- GHC 8.10, 9.0, 9.2, and their package sets have been removed.
+
+- GHC 8.6 and its package set have been removed. It was only used to bootstrap GHC for powerpc64le, but this was probably broken anyway.
```
Done. I also rearranged the release notes a bit accordingly.

emilazy commented:
Everything good to move forward with merging this and https://github.com/NixOS/nixpkgs/pull/440271 so I can rebase https://github.com/NixOS/nixpkgs/pull/440273, then?
